phoenix-core-client/src/main/java/org/apache/phoenix/compile/WhereOptimizer.java

```diff
@@ -1945,6 +1948,150 @@ public KeySlots visitLeave(InListExpression node, List<KeySlots> childParts) {
+        keyExpressions.add(lhs);
+
+      } else if (lhs instanceof ArrayElemRefExpression && rhs instanceof RowKeyColumnExpression) {
+        ArrayElemRefExpression arrayElemRefExpr = (ArrayElemRefExpression) lhs;
+        if (!(arrayElemRefExpr.getChildren().get(0) instanceof LiteralExpression)) {
+          return false;
+        }
+        // Capture {@link RowKeyColumnExpression} for the generation of key slots.
+        keyExpressions.add(rhs);
+      }
+      return true;
+    }
+
+    @Override
+    public Iterator<Expression> visitEnter(ArrayAnyComparisonExpression node) {
+      ArrayList<Expression> keyExpressions = new ArrayList<>();
+      if (prepareForArrayAnyComparisonKeySlots(node, keyExpressions)) {
+        return keyExpressions.iterator();
+      }
+      // If the scan ranges cannot be extracted, we return an empty iterator
+      return Collections.emptyIterator();
+    }
+
+    @Override
+    public KeySlots visitLeave(ArrayAnyComparisonExpression node, List<KeySlots> childParts) {
+      if (childParts == null || childParts.isEmpty()) {
+        return null;
+      }
+      // Doing type casting is safe here as we won't have reached here unless the expression
+      // tree is of the form expected by the method prepareForArrayAnyComparisonKeySlots.
+      Expression arrayExpr = node.getChildren().get(0);
+      PhoenixArray arr = (PhoenixArray) ((LiteralExpression) arrayExpr).getValue();
+      int numElements = arr.getDimensions();
+
+      ComparisonExpression comparisonExpr = (ComparisonExpression) node.getChildren().get(1);
+      Expression lhsExpr = comparisonExpr.getChildren().get(0);
+      Expression rhsExpr = comparisonExpr.getChildren().get(1);
+      ArrayElemRefExpression arrayElemRefExpr;
+      if (lhsExpr instanceof ArrayElemRefExpression) {
+        arrayElemRefExpr = (ArrayElemRefExpression) lhsExpr;
+      } else {
+        arrayElemRefExpr = (ArrayElemRefExpression) rhsExpr;
+      }
+
+      KeySlots childSlots = childParts.get(0);
+      KeySlot childSlot = childSlots.getSlots().get(0);
+      KeyPart childPart = childSlot.getKeyPart();
+      PColumn column = childPart.getColumn();
+
+      List<KeyRange> keyRanges = new ArrayList<>();
+      try {
+        Expression coerceExpr = CoerceExpression.create(arrayElemRefExpr, column.getDataType(),
+          column.getSortOrder(), column.getMaxLength());
+        for (int i = 1; i <= numElements; i++) {
+          arrayElemRefExpr.setIndex(i);
+          KeyRange keyRange = childPart.getKeyRange(CompareOperator.EQUAL, coerceExpr);
+          if (
+            keyRange == null || keyRange == KeyRange.EMPTY_RANGE
+              || keyRange == KeyRange.IS_NULL_RANGE
+          ) {
+            // Skip null range along with empty range as null check is done via IS NULL as
+            // per SQL standards
+            continue;
+          }
+          keyRanges.add(keyRange);
+        }
+      } catch (Exception e) {
+        LOGGER.warn(String.format(
```

Comment:
Let's avoid using `String.format()` for loggers, we can use parameterized log format with `{}` for each object that needs to be logged